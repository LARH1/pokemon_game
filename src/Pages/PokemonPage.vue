<template>
<div>
    <div v-if="pokemon.id">
        <poke-image :pokeId="pokemon.id" :isVisible="showPoke" />
        <poke-options :options="poke_data" @poke-selection="CheckSelected" />

        <div v-if="showAnswer" class="fade-in">
            <h3>{{message}}</h3>
            <button class="btn-restart" @click="Restart">Reiniciar</button>
            <br>
            <br>
        </div>
    </div>
    <h2 v-else>Cargando...</h2>
</div>
</template>

<script>
import getOptions from "@/helpers/options";
import PokeImage from "@/components/PokeImage.vue";
import PokeOptions from "@/components/PokeOptions.vue";
export default
{
    components:
    {
        "poke-image": PokeImage,
        "poke-options": PokeOptions,
    },
    data()
    {
        return {
            id: 1,
            poke_data: [],
            pokemon:
            {},
            showPoke: false,
            showAnswer: false,
            message: "",
        }
    },
    methods:
    {
        /**
         * Obtener la lista de pokemones
         */
        async GetData()
        {
            // Obtener las opciones
            this.poke_data = await getOptions();
            // Obtener el ganador
            const id = Math.floor(Math.random() * 4);
            this.pokemon = this.poke_data[id];
        },

        /**
         * Comprobar el pokemon seleccionado
         */
        CheckSelected(pokeId)
        {
            this.showAnswer = true;
            this.showPoke = true;
            if (pokeId == this.pokemon.id)
                this.message = `!Correcto!, es ${this.pokemon.name}`;
            else
                this.message = `Oops!, era ${this.pokemon.name}`;
        },

        /**
         * Reiniciar juego
         */
        Restart()
        {
            this.poke_data = [];
            this.pokemon = {};
            this.message = "";
            this.showPoke = false;
            this.showAnswer = false;
            this.GetData();
        }
    },
    mounted()
    {
        this.GetData();
    }
}
</script>

<style>

</style>
