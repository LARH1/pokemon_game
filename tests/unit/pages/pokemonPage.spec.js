import PokemonPage from "@/Pages/PokemonPage";
import { shallowMount } from "@vue/test-utils";

describe("PokemonPage",() =>
{
    // Montar al inicio. Se puede reutilizar
    let wrap;
    beforeEach(() =>
    {
        wrap = shallowMount(PokemonPage);
    });

    test("Debe coincidir con el snapshot cuando cargó la info",() =>
    {
        const wrap1 = shallowMount(PokemonPage,{
            data()
            {
                return {
                    id: 1,
                    poke_data: [],
                    pokemon:
                    {
                        id: 1,
                        name: "BULBASAUR"
                    },
                    showPoke: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })
        expect(wrap1.html()).toMatchSnapshot();
    });

    test("Debe cargar la data en el mounted()",() =>
    {
        // Spy del GetData
        const getDataSpy = jest.spyOn(PokemonPage.methods,"GetData");
        // Montar Page
        const wrp2 = shallowMount(PokemonPage);
        expect(getDataSpy).toHaveBeenCalled(); // Se llamó al montar
    });

    test("PokemonImage y PokeOptions deben existir",() =>
    {
        // crear wrap
        const wrap1 = shallowMount(PokemonPage,{
            data()
            {
                return {
                    id: 1,
                    poke_data: [],
                    pokemon:
                    {
                        id: 1,
                        name: "BULBASAUR"
                    },
                    showPoke: false,
                    showAnswer: false,
                    message: "",
                }
            }
        });
        // compomente  poke-image (-stub???)
        const image = wrap1.find("poke-image-stub");
        expect(image.exists()).toBeTruthy();
        // compomente  poke-options (-stub???)
        const options = wrap1.find("poke-options-stub");
        expect(options.exists()).toBeTruthy();
        // PokeImage :pokeId=1
        expect(image.attributes("pokeid")).toBe("1")
        // PokeOptions :options debe existir
        expect(options.attributes("options")).toBe("");
    });

    test("PokemonImage y PokeOptions deben existir",async () =>
    {
        // crear wrap
        const wrap = shallowMount(PokemonPage,{
            data()
            {
                return {
                    id: 1,
                    poke_data: [],
                    pokemon:
                    {
                        id: 1,
                        name: "BULBASAUR"
                    },
                    showPoke: false,
                    showAnswer: false,
                    message: "",
                }
            }
        });
        // Comprobar el poke 1
        await wrap.vm.CheckSelected(1);
        expect(wrap.find("h3").exists()).toBe(true);
        expect(wrap.find("h3").text()).toBe("¡Correcto, es BULBASAUR!");
        expect(wrap.vm.showPoke).toBe(true);

        // Comprobar el poke != 1
        await wrap.vm.CheckSelected(2);
        expect(wrap.find("h3").exists()).toBe(true);
        expect(wrap.vm.message).toBe("Oops!, era BULBASAUR");
        expect(wrap.vm.showPoke).toBe(true);
    })
})