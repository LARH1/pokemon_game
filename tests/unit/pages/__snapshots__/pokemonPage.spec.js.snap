// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Debe coincidir con el snapshot cuando carg√≥ la info 1`] = `
<div>
  <div>
    <poke-image-stub pokeid="1" isvisible="false"></poke-image-stub>
    <poke-options-stub options=""></poke-options-stub>
    <!--v-if-->
  </div>
</div>
`;
